File: playground/server/modules/sessions.ts
import type { Express, Request, Response as ExpressResponse } from 'express'
import { createSolanaRpc, type KeyPairSigner } from '@solana/kit'
import { Mppx, session, createMemorySessionStore, type SessionStore } from '@solana/mpp/server'
import { toWebRequest, logPayment } from '../shared/utils.js'
import { USDC_DECIMALS, USDC_MINT } from '../shared/constants.js'

const Web = globalThis

interface RegisterOptions {
  recipient: string
  network: string
  secretKey: string
  feePayerSigner: KeyPairSigner
  rpcUrl: string
}

const TOKEN_CHUNKS = [
  'A payment channel ',
  'lets a client and server ',
  'authorize many small ',
  'off-chain debits ',
  'against a single on-chain ',
  'deposit, settling the highest ',
  'cumulative voucher at close.',
]

/**
 * Register the two session-gated demo endpoints using the in-process
 * `session()` MPP method. Replaces the prior `pay` CLI sidecar reverse-proxy
 * with a fully in-process payment-channel session implementation.
 *
 * Routes:
 * - GET  /sessions/stream  — pay-per-chunk SSE, cap 1.00 USDC, 0.0001 USDC/chunk
 * - POST /sessions/stream  — voucher commits for the stream endpoint
 * - POST /sessions/compute — pay-per-call compute, cap 0.50 USDC, 0.005 USDC/call
 *                            (also accepts voucher commits)
 * - POST /__402/session/deliveries — SessionFetchClient's delivery reservation
 * - POST /__402/session/commit     — body-voucher commit variant of the above
 * - GET  /sessions/receipt/:channelId — settle-status poll for the UI
 */
export function registerSessions(app: Express, opts: RegisterOptions): void {
  const { recipient, network, secretKey, feePayerSigner, rpcUrl } = opts

  // Shared store across all session() methods so `/sessions/receipt/:channelId`
  // can read the settled signature whichever endpoint opened the channel.
  const sharedStore: SessionStore = createMemorySessionStore()

  // The session() idle-close watchdog and the `close` action only settle
  // on-chain when both a merchant signer AND a kit rpc client are configured.
  const rpc = createSolanaRpc(rpcUrl)

  // ── /sessions/stream — pay-per-chunk SSE ──
  const streamMethod = session({
    cap: 1_000_000n, // 1.00 USDC
    closeDelayMs: 2_000, // settle ~2s after the stream ends so the UI can poll quickly
    currency: USDC_MINT,
    decimals: USDC_DECIMALS,
    // Real on-chain opens: the browser pre-signs a payment-channel open
    // transaction (fee payer = operator) and the server completes the
    // signature, broadcasts, and waits for confirmation before metering.
    modes: ['pull'],
    network,
    openTxSubmitter: 'server',
    operator: feePayerSigner.address,
    paymentChannelPayerSigner: feePayerSigner,
    pricing: { perDelivery: 100n }, // 0.0001 USDC per chunk
    pullVoucherStrategy: 'clientVoucher',
    recipient,
    rpc,
    rpcUrl,
    signer: feePayerSigner,
    store: sharedStore,
  })

  const streamMppx = Mppx.create({
    methods: [streamMethod],
    secretKey,
  }) as unknown as MppxLike

  const streamHandler = streamMppx.session({
    cap: '1000000',
    currency: USDC_MINT,
    description: 'Metered token stream',
  })

  app.get('/sessions/stream', async (req: Request, res: ExpressResponse) => {
    const result = await streamHandler(toWebRequest(req))

    if (result.status === 402) {
      const challenge = result.challenge as globalThis.Response
      res.writeHead(challenge.status, Object.fromEntries(challenge.headers))
      res.end(await challenge.text())
      return
    }

    // Stream tokens as SSE; each chunk costs 0.0001 USDC.
    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      Connection: 'keep-alive',
    })
    for (const chunk of TOKEN_CHUNKS) {
      res.write(`data: ${JSON.stringify({ chunk, cost: '100' })}\n\n`)
      await new Promise(r => setTimeout(r, 80))
    }
    res.write('data: [DONE]\n\n')
    res.end()
  })

  // Voucher commits: SessionFetchClient POSTs each commit back to the URL it
  // opened the session against, with the signed voucher in the Authorization
  // credential. The method's verify path applies it; the body is just an ack.
  app.post('/sessions/stream', async (req: Request, res: ExpressResponse) => {
    const result = await streamHandler(toWebRequest(req))
    if (result.status === 402) {
      const challenge = result.challenge as globalThis.Response
      res.writeHead(challenge.status, Object.fromEntries(challenge.headers))
      res.end(await challenge.text())
      return
    }
    await sendWebResponse(res, result.withReceipt(commitAck(req)) as globalThis.Response)
  })

  // ── /sessions/compute — pay-per-call compute ──
  const computeMethod = session({
    cap: 500_000n, // 0.50 USDC
    closeDelayMs: 2_000,
    currency: USDC_MINT,
    decimals: USDC_DECIMALS,
    modes: ['pull'],
    network,
    openTxSubmitter: 'server',
    operator: feePayerSigner.address,
    paymentChannelPayerSigner: feePayerSigner,
    pricing: { perDelivery: 5_000n }, // 0.005 USDC per call
    pullVoucherStrategy: 'clientVoucher',
    recipient,
    rpc,
    rpcUrl,
    signer: feePayerSigner,
    store: sharedStore,
  })

  const computeMppx = Mppx.create({
    methods: [computeMethod],
    secretKey,
  }) as unknown as MppxLike

  app.post('/sessions/compute', async (req: Request, res: ExpressResponse) => {
    const result = await computeMppx.session({
      cap: '500000',
      currency: USDC_MINT,
      description: 'Voucher-billed inference call',
    })(toWebRequest(req))

    if (result.status === 402) {
      const challenge = result.challenge as globalThis.Response
      res.writeHead(challenge.status, Object.fromEntries(challenge.headers))
      res.end(await challenge.text())
      return
    }

    const body = (req.body ?? {}) as { prompt?: string; deliveryId?: string }
    // Voucher commits arrive on the same URL the session was opened against
    // (see the /sessions/stream POST handler above) — ack instead of computing.
    if (body.deliveryId) {
      await sendWebResponse(res, result.withReceipt(commitAck(req)) as globalThis.Response)
      return
    }

    const prompt = body.prompt ?? ''
    const response = result.withReceipt(
      Web.Response.json({
        prompt,
        output: `Echo: ${prompt} (computed for 0.005 USDC)`,
        computedAt: new Date().toISOString(),
      }),
    ) as globalThis.Response
    logPayment(req.path, response)
    res.writeHead(response.status, Object.fromEntries(response.headers))
    res.end(await response.text())
  })

  // ── Side-channel metering routes ──
  // SessionFetchClient reserves capacity for each metered delivery at
  // /__402/session/deliveries before signing + committing the voucher.
  const sessionRoutes = session.routes({
    cap: 1_000_000n,
    currency: USDC_MINT,
    decimals: USDC_DECIMALS,
    network,
    operator: feePayerSigner.address,
    pricing: { perDelivery: 100n },
    recipient,
    store: sharedStore,
  })

  app.post('/__402/session/deliveries', async (req: Request, res: ExpressResponse) => {
    await sendWebResponse(res, await sessionRoutes.deliveries(toWebRequest(req)))
  })
  app.post('/__402/session/commit', async (req: Request, res: ExpressResponse) => {
    await sendWebResponse(res, await sessionRoutes.commit(toWebRequest(req)))
  })

  // Receipt poll endpoint: the UI hits this after the stream ends to learn
  // the on-chain settle signature. The idle-close watchdog fires
  // ~closeDelayMs after the last voucher and, with `signer` + `rpc`
  // configured above, attempts the on-chain settle-and-distribute.
  app.get('/sessions/receipt/:channelId', async (req: Request, res: ExpressResponse) => {
    const channelIdParam: unknown = req.params.channelId
    if (typeof channelIdParam !== 'string' || channelIdParam.length === 0) {
      return res.status(400).json({ error: 'invalid-channel-id' })
    }
    const state = await sharedStore.getChannel(channelIdParam)
    if (!state) return res.status(404).json({ error: 'channel-not-found' })
    return res.json({
      channelId: state.channelId,
      cumulative: state.cumulative.toString(),
      deposit: state.deposit.toString(),
      finalized: state.finalized,
      settledSignature: state.settledSignature ?? null,
    })
  })
}

/** Minimal JSON ack for a voucher commit — mirrors the CommitReceipt shape
 * the SessionFetchClient parses from the commit response. */
function commitAck(req: Request): globalThis.Response {
  const body = (req.body ?? {}) as { amount?: string; deliveryId?: string }
  return Web.Response.json({
    amount: body.amount ?? '0',
    deliveryId: body.deliveryId ?? '',
    status: 'committed',
  })
}

async function sendWebResponse(res: ExpressResponse, response: globalThis.Response): Promise<void> {
  res.writeHead(response.status, Object.fromEntries(response.headers))
  res.end(await response.text())
}

type MppxLike = {
  session: (params: { amount?: string; cap?: string; currency: string; description: string }) => (
    request: globalThis.Request,
  ) => Promise<{
    status: number
    challenge?: globalThis.Response
    withReceipt: (r: globalThis.Response) => globalThis.Response
  }>
}

File: go/examples/playground-api/sessions.go
package main

// Two session-gated demo endpoints driven by the in-process session method,
// the reserve/commit metering side channel, and the settle-status receipt
// poll. Both methods share one channel store so the receipt endpoint can
// read the settled signature whichever endpoint opened the channel.

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/solana-foundation/pay-kit/go/paycore"
	"github.com/solana-foundation/pay-kit/go/protocols/mpp/intents"
	server "github.com/solana-foundation/pay-kit/go/protocols/mpp/server"
)

// tokenChunks is the canned token stream payload.
var tokenChunks = []string{
	"A payment channel ",
	"lets a client and server ",
	"authorize many small ",
	"off-chain debits ",
	"against a single on-chain ",
	"deposit, settling the highest ",
	"cumulative voucher at close.",
}

// registerSessions mounts the session endpoints and returns the watchdog
// shutdown hook.
//
// Routes:
//   - GET  /sessions/stream: pay-per-chunk SSE, cap 1.00 USDC, 0.0001 USDC/chunk
//   - GET  /api/v1/stream: TS-reference alias for /sessions/stream
//   - POST /sessions/stream: voucher commits for the stream endpoint
//   - POST /api/v1/stream: TS-reference alias for voucher commits
//   - POST /sessions/compute: pay-per-call compute, cap 0.50 USDC, 0.005 USDC/call
//     (also accepts voucher commits)
//   - POST /__402/session/deliveries: SessionFetch-style delivery reservation
//   - POST /__402/session/commit: body-voucher commit variant of the above
//   - GET  /sessions/receipt/{channelId}: settle-status poll for the UI
func registerSessions(mux *http.ServeMux, a *app) (func(), error) {
	// Shared store across both session methods so /sessions/receipt can read
	// channel state regardless of which endpoint opened the channel.
	sharedStore := server.NewMemoryChannelStore()
	strategy := intents.SessionPullVoucherStrategyClientVoucher

	newMethod := func(cap uint64) (*server.Session, error) {
		return server.NewSession(server.SessionOptions{
			Operator:  a.feePayer.PublicKey().String(),
			Recipient: a.recipient,
			Cap:       cap,
			Currency:  paycore.USDCMainnetMint,
			Decimals:  usdcDecimals,
			Network:   a.network,
			SecretKey: a.secretKey,
			// Real on-chain opens: the browser pre-signs a payment-channel
			// open transaction (fee payer = operator) and the server
			// completes the signature, broadcasts, and waits for
			// confirmation before metering.
			Modes:               []intents.SessionMode{intents.SessionModePull},
			PullVoucherStrategy: &strategy,
			OpenTxSubmitter:     server.OpenTxSubmitterServer,
			// Settle roughly two seconds after the stream ends so the UI's
			// receipt poll resolves quickly.
			CloseDelay:                2 * time.Second,
			PaymentChannelPayerSigner: a.feePayer,
			Signer:                    a.feePayer,
			RPC:                       a.rpcClient,
			Store:                     sharedStore,
		})
	}

	streamSession, err := newMethod(1_000_000) // 1.00 USDC
	if err != nil {
		return nil, fmt.Errorf("stream session method: %w", err)
	}
	computeSession, err := newMethod(500_000) // 0.50 USDC
	if err != nil {
		streamSession.Shutdown()
		return nil, fmt.Errorf("compute session method: %w", err)
	}
	shutdown := func() {
		streamSession.Shutdown()
		computeSession.Shutdown()
	}

	streamGate := server.SessionMiddleware(streamSession, func(*http.Request) (server.SessionChallengeOptions, error) {
		return server.SessionChallengeOptions{Cap: "1000000", Description: "Metered token stream"}, nil
	})
	computeGate := server.SessionMiddleware(computeSession, func(*http.Request) (server.SessionChallengeOptions, error) {
		return server.SessionChallengeOptions{Cap: "500000", Description: "Voucher-billed inference call"}, nil
	})

	streamHandler := streamGate(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		stream := server.NewMeteredStream(w)
		w.WriteHeader(http.StatusOK)
		for _, chunk := range tokenChunks {
			if err := stream.WriteJSON(map[string]string{"chunk": chunk, "cost": "100"}); err != nil {
				return
			}
			time.Sleep(80 * time.Millisecond)
		}
		_ = stream.WriteDone()
	}))
	// GET /sessions/stream and /api/v1/stream: stream tokens as SSE; each
	// chunk costs 0.0001 USDC.
	mux.Handle("GET /sessions/stream", streamHandler)
	mux.Handle("GET /api/v1/stream", streamHandler)

	streamCommitHandler := streamGate(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, http.StatusOK, commitAck(r))
	}))
	// POST /sessions/stream and /api/v1/stream: voucher commits arrive on
	// the URL the session was opened against, with the signed voucher in the
	// Authorization credential. The middleware's verify path applies it; the
	// body is an ack.
	mux.Handle("POST /sessions/stream", streamCommitHandler)
	mux.Handle("POST /api/v1/stream", streamCommitHandler)

	// POST /sessions/compute: pay-per-call compute; the same handler also
	// accepts voucher commits (a deliveryId in the body discriminates).
	mux.Handle("POST /sessions/compute", computeGate(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var body struct {
			Prompt     string `json:"prompt"`
			DeliveryID string `json:"deliveryId"`
		}
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body.DeliveryID != "" {
			writeJSON(w, http.StatusOK, map[string]string{
				"amount":     "0",
				"deliveryId": body.DeliveryID,
				"status":     "committed",
			})
			return
		}
		logPayment(r.URL.Path, w.Header())
		writeJSON(w, http.StatusOK, map[string]string{
			"prompt":     body.Prompt,
			"output":     "Echo: " + body.Prompt + " (computed for 0.005 USDC)",
			"computedAt": time.Now().UTC().Format(time.RFC3339),
		})
	})))

	// Side-channel metering routes: SessionFetch-style clients reserve
	// capacity for each metered delivery before signing + committing the
	// voucher. Both handlers share the methods' channel store.
	routes := streamSession.Routes()
	mux.HandleFunc("POST /__402/session/deliveries", routes.Deliveries)
	mux.HandleFunc("POST /__402/session/commit", routes.Commit)

	// Receipt poll endpoint: the UI hits this after the stream ends to learn
	// the on-chain settle signature. The idle-close watchdog fires about
	// CloseDelay after the last voucher and, with Signer + RPC configured
	// above, attempts the on-chain settle-and-distribute.
	mux.HandleFunc("GET /sessions/receipt/{channelId}", func(w http.ResponseWriter, r *http.Request) {
		channelID := r.PathValue("channelId")
		if channelID == "" {
			writeJSONError(w, http.StatusBadRequest, "invalid-channel-id")
			return
		}
		state, err := sharedStore.GetChannel(r.Context(), channelID)
		if err != nil || state == nil {
			writeJSONError(w, http.StatusNotFound, "channel-not-found")
			return
		}
		var settledSignature any
		if state.SettledSignature != nil {
			settledSignature = *state.SettledSignature
		}
		writeJSON(w, http.StatusOK, map[string]any{
			"channelId":        state.ChannelID,
			"cumulative":       fmt.Sprintf("%d", state.Cumulative),
			"deposit":          fmt.Sprintf("%d", state.Deposit),
			"finalized":        state.Finalized,
			"settledSignature": settledSignature,
		})
	})

	return shutdown, nil
}

// commitAck is the minimal CommitReceipt-shaped JSON ack the stream commit
// handler returns.
func commitAck(r *http.Request) map[string]string {
	var body struct {
		Amount     string `json:"amount"`
		DeliveryID string `json:"deliveryId"`
	}
	_ = json.NewDecoder(r.Body).Decode(&body)
	if body.Amount == "" {
		body.Amount = "0"
	}
	return map[string]string{
		"amount":     body.Amount,
		"deliveryId": body.DeliveryID,
		"status":     "committed",
	}
}
